diff-test/core/src/main.ts

```diff
@@ -0,0 +1,437 @@
+import * as fs from 'fs';
+import _ from 'underscore';
+import timeSpan from 'time-span';
+import cloneDeep from 'clone-deep';
+import {
+  BlockHistory,
+  stakingWithoutSlashing,
+  bondBasedConsumerVotingPower
+} from './properties.js';
+import { Sanity as SanityChecker } from './sanity.js';
+import { Model } from './model.js';
+import {
+  createSmallSubsetOfCoveringTraces,
+  dumpTrace,
+  forceMakeEmptyDir,
+  logEventData,
+} from './util.js';
+import {
+  P,
+  C,
+  NUM_VALIDATORS,
+  BLOCK_SECONDS,
+  TOKEN_SCALAR,
+  MAX_BLOCK_ADVANCES,
+} from './constants.js';
+
+interface Action {
+  kind: string;
+}
+
+type Delegate = {
+  kind: string;
+  val: number;
+  amt: number;
+};
+
+type Undelegate = {
+  kind: string;
+  val: number;
+  amt: number;
+};
+
+type JumpNBlocks = {
+  kind: string;
+  chains: string[];
+  n: number;
+  secondsPerBlock: number;
+};
+
+type Deliver = {
+  kind: string;
+  chain: string;
+  numPackets: string;
+};
+
+type ConsumerSlash = {
+  kind: string;
+  val: number;
+  infractionHeight: number;
+  isDowntime: number;
+};
+
+/**
+ * Takes an object where values are probabilities and returns a random
+ * key based on the distribution.
+ */
+function weightedRandomKey(distr) {
+  const scalar = _.reduce(_.values(distr), (sum, y) => sum + y, 0);
```

Comment:
JS has a built in reduce. In general, you probably shouldn't need to import underscore anywhere in this project. It is a relic from the days (10+ years ago) when the js stdlib was very minimal.

Reply:
Also Object.values()

Reply:
> JS has a built in reduce. In general, you probably shouldn't need to import underscore anywhere in this project. It is a relic from the days (10+ years ago) when the js stdlib was very minimal.

Ok, I'll make a pass to see if I can do things in vanilla. I got the impression underscore is really quite helpful though.

Reply:
Already found one example where underscore seems to save writing a lot of code

https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_pick

Reply:
I do take your point though, which is that I've probably overused it.